pkg/cloudprovider/providers/aws/retry_handler.go

```diff
@@ -0,0 +1,137 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/golang/glog"
+)
+
+const (
+	decayIntervalSeconds = 30
+	decayFraction        = 0.8
+	maxDelay             = 60 * time.Second
+)
+
+// CrossRequestRetryDelay inserts delays before AWS calls, when we are observing RequestLimitExceeded errors
+// Note that we share a CrossRequestRetryDelay across multiple AWS requests; this is a process-wide back-off,
+// whereas the aws-sdk-go implements a per-request exponential backoff/retry
+type CrossRequestRetryDelay struct {
+	mutex sync.Mutex
+
+	// We count all requests and the number of requests which hit a
+	// RequestLimit We only really care about 'recent' requests, so we
+	// decay the counts exponentially to bias towards recent values.
+	countErrorsRequestLimit float32
+	countRequests           float32
+	lastDecay               int64
+}
+
+// Create a new CrossRequestRetryDelay
+func NewCrossRequestRetryDelay() *CrossRequestRetryDelay {
+	return &CrossRequestRetryDelay{
+		lastDecay: time.Now().Unix(),
+		// Bias so that if the first request hits the limit we don't immediately apply the full delay
+		countRequests: 4,
+	}
+}
+
+// Added to the Sign chain; called before each request
+func (c *CrossRequestRetryDelay) BeforeSign(r *request.Request) {
+	now := time.Now()
+	delay := c.computeDelay(now, r)
+	if delay > 0 {
+		glog.Warningf("Inserting delay before AWS request (%s) to avoid RequestLimitExceeded: %s",
+			describeRequest(r), delay.String())
+		r.Config.SleepDelay(delay)
+
+		// Avoid clock skew problems
+		r.Time = now
+	}
+}
+
+// Return a user-friendly string describing the request, for use in log messages
+func describeRequest(r *request.Request) string {
+	service := r.ClientInfo.ServiceName
+
+	name := "?"
+	if r.Operation != nil {
+		name = r.Operation.Name
+	}
+
+	return service + "::" + name
+}
+
+// Added to the AfterRetry chain; called after any error
+func (c *CrossRequestRetryDelay) AfterRetry(r *request.Request) {
+	if r.Error != nil {
+		if awsError, ok := r.Error.(awserr.Error); ok {
+			if awsError.Code() == "RequestLimitExceeded" {
+				c.mutex.Lock()
+				c.countErrorsRequestLimit += 1.0
+				c.mutex.Unlock()
+
+				glog.Warningf("Got RequestLimitExceeded error on AWS request (%s)",
+					describeRequest(r))
+			}
+		}
+	}
+}
+
+// Computes the delay required for a request, also updating internal state for this request
+func (c *CrossRequestRetryDelay) computeDelay(now time.Time, r *request.Request) time.Duration {
+	c.mutex.Lock()
+
+	// Apply exponential decay to the counters
+	timeDeltaSeconds := now.Unix() - c.lastDecay
```

Comment:
I added a set of simple unit tests that test the "macro" behaviour e.g. no errors => no delay; errors => delay with no flapping; errors go close to max eventually; recovers from errors eventually.

There wasn't a need to mock the clock, because computeDelay takes now as an argument anyway.